### PR TITLE
Feat/study data flow

### PR DESCRIPTION
**Pull request by fung-hei-man**
## Description
- Add Redux Actions

action          | 說明  | 
--------------|:-----:|
SET_STUDY_FILTER     | 設定study文章的篩選統計條件
CLEAR_STUDY_FILTER  | 清除所有study文章的篩選統計條件
INIT_STUDY_STAT  | 從backend取得統計&分類頂目並加到store
ADD_STUDY_STAT     | 新增統計&分類項目(後台)
UPDATE_STUDY_STAT     | 更新已存在的統計&分類項目(後台)
DELETE_STUDY_STAT     | 刪除已存在的統計&分類項目(後台)

- Store selected filter options in Redux
- Add badges of selected filter options
- Use context to control filter status (opened, editing, etc.)

## Github issue
- https://github.com/JiaAnTW/NCKU_Trans/issues/47
## Scope
- /
- /admin/study

### CODE REVIEW

File: resources/assets/js/page/Study/useInitData.js [hunk 1]
```diff
@@ -2,11 +2,17 @@ import { useEffect } from 'react';
...
+import { fetchStudyStat, fetchStudyType } from '../../model/middleware/study';
 
 function useInitData({ isAdmin, num }) {
     const dispatch = useDispatch();
     const isFinishRequest = useRequest();
 
+    useEffect(() => {
+        dispatch(fetchStudyType());
+        dispatch(fetchStudyStat());
+    }, []);
```
Comment (JiaAnTW) [inline resources/assets/js/page/Study/useInitData.js:14]:
```js
    useEffect(() => {
        dispatch(fetchStudyType());
        dispatch(fetchStudyStat());
    }, [dispatch]);
```

File: resources/assets/js/page/Study/Filter/ItemFilter/ItemFilterColumn/index.js [hunk 2]
```diff
@@ -7,18 +8,44 @@ import FormControlLabel from '@material-ui/core/FormControlLabel';
...
 
 function ItemFilterColumn({ optionsArr }) {
     const classes = useStyles();
+    const dispatch = useDispatch();
+    const selectedFilter = useSelector(selectedFilterSelector);
+
+    const onChangeCheckbox = (id, checked) => {
+        dispatch({
+            type: SET_STUDY_FILTER,
+            payload: { tagType: optionsArr.type, tagId: id, checked },
```
Comment (JiaAnTW) [inline resources/assets/js/page/Study/Filter/ItemFilter/ItemFilterColumn/index.js:22]: 這裡的optionArr感覺用參數傳進來? 讓他變成接近pure function

File: resources/assets/js/page/Study/useInitData.js (hunk 1 above)
Comment (fung-hei-man) [reply resources/assets/js/page/Study/useInitData.js:14]: 啊啊啊stash code之後忘記加回去sorry

**Comment by fung-hei-man** [comment]
@JiaAnTW Fixed above 2 issues. Please check again, thanks!

**Review by JiaAnTW — APPROVED** [review]
LGTM 👍

**Pull request merged**